### PR TITLE
Reserve memory before projecting rows

### PR DESCRIPTION
**Pull request by yingsu00**
Previously the BlockBuilders always grow their internal arrays from 1
element, and call growCapacity() when adding new elements. growCapacity()
is every expensive in both CPU and memory allocations. Sometimes we
know the number of elements that are going to be added in a batch, and
we can pre-reserved the memory to avoid repeating memory growth and
copying.

BenchmarkPageProcessor shows 33% improvement in projection's throughput
(higher is better) and 4x reduction in gc.alloc.rate.norm
(lower is better) in GC benchmarks.

Before
```
Benchmark                                                         (projectionDataType)   Mode  Cnt       Score        Error   Units
BenchmarkPageProcessor.compiled                                                 BIGINT  thrpt   10    8374.449 ±   1023.859   ops/s
BenchmarkPageProcessor.compiled:·gc.alloc.rate                                  BIGINT  thrpt   10    1451.783 ±    179.079  MB/sec
BenchmarkPageProcessor.compiled:·gc.alloc.rate.norm                             BIGINT  thrpt   10  363640.395 ±      1.385    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space                         BIGINT  thrpt   10    1448.711 ±   1904.852  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space.norm                    BIGINT  thrpt   10  368448.852 ± 487748.838    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space                     BIGINT  thrpt   10       0.019 ±      0.054  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space.norm                BIGINT  thrpt   10       4.904 ±     13.838    B/op
BenchmarkPageProcessor.compiled:·gc.count                                       BIGINT  thrpt   10       6.000               counts
BenchmarkPageProcessor.compiled:·gc.time                                        BIGINT  thrpt   10      27.000                   ms
BenchmarkPageProcessor.compiled                                                 DOUBLE  thrpt   10    8805.494 ±    916.612   ops/s
BenchmarkPageProcessor.compiled:·gc.alloc.rate                                  DOUBLE  thrpt   10    1527.888 ±    159.617  MB/sec
BenchmarkPageProcessor.compiled:·gc.alloc.rate.norm                             DOUBLE  thrpt   10  363640.100 ±      0.011    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space                         DOUBLE  thrpt   10    1716.294 ±   1812.190  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space.norm                    DOUBLE  thrpt   10  414368.217 ± 436717.530    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space                     DOUBLE  thrpt   10       0.008 ±      0.023  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space.norm                DOUBLE  thrpt   10       1.942 ±      5.771    B/op
BenchmarkPageProcessor.compiled:·gc.count                                       DOUBLE  thrpt   10       7.000               counts
BenchmarkPageProcessor.compiled:·gc.time                                        DOUBLE  thrpt   10      30.000                   ms
```
After
```
Benchmark                                                         (projectionDataType)   Mode  Cnt       Score        Error   Units
BenchmarkPageProcessor.compiled                                                 BIGINT  thrpt   10   11140.511 ±   1245.100   ops/s
BenchmarkPageProcessor.compiled:·gc.alloc.rate                                  BIGINT  thrpt   10     494.381 ±     55.174  MB/sec
BenchmarkPageProcessor.compiled:·gc.alloc.rate.norm                             BIGINT  thrpt   10   93152.321 ±      1.149    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space                         BIGINT  thrpt   10     534.785 ±    705.834  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space.norm                    BIGINT  thrpt   10  105433.423 ± 139043.448    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space                     BIGINT  thrpt   10       0.012 ±      0.045  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space.norm                BIGINT  thrpt   10       2.275 ±      8.238    B/op
BenchmarkPageProcessor.compiled:·gc.count                                       BIGINT  thrpt   10       6.000               counts
BenchmarkPageProcessor.compiled:·gc.time                                        BIGINT  thrpt   10      27.000                   ms
BenchmarkPageProcessor.compiled                                                 DOUBLE  thrpt   10   11050.149 ±   1224.166   ops/s
BenchmarkPageProcessor.compiled:·gc.alloc.rate                                  DOUBLE  thrpt   10     491.177 ±     54.807  MB/sec
BenchmarkPageProcessor.compiled:·gc.alloc.rate.norm                             DOUBLE  thrpt   10   93152.079 ±      0.009    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space                         DOUBLE  thrpt   10     365.061 ±    889.401  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Eden_Space.norm                    DOUBLE  thrpt   10   68525.405 ± 167416.354    B/op
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space                     DOUBLE  thrpt   10       0.006 ±      0.020  MB/sec
BenchmarkPageProcessor.compiled:·gc.churn.PS_Survivor_Space.norm                DOUBLE  thrpt   10       1.133 ±      3.612    B/op
BenchmarkPageProcessor.compiled:·gc.count                                       DOUBLE  thrpt   10       3.000               counts
BenchmarkPageProcessor.compiled:·gc.time                                        DOUBLE  thrpt   10      13.000                   ms
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-common/src/main/java/com/facebook/presto/common/block/MapBlockBuilder.java [hunk 1]
```diff
@@ -470,6 +471,15 @@ public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus)
...
     }
 
+    @Override
+    public void reserve(int positionCount)
+    {
+        assert positionCount > 0;
+
+        mapIsNull = ensureCapacity(mapIsNull, positionCount);
+        offsets = ensureCapacity(offsets, positionCount + 1);
+    }
```
Comment (sujay-jain) [inline presto-common/src/main/java/com/facebook/presto/common/block/MapBlockBuilder.java:481]: curious why we don't ```updateDataSize``` here? It is coz the primitive types take care of it for complex types?

**Review by sujay-jain — APPROVED** [review]
LGTM. Thanks for adding the tests

**Comment by yingsu00** [comment]
Thanks @sujay-jain. @mbasmanova Will you please to take another look? This PR adds a new function to BlockBuilders.

File: presto-common/src/main/java/com/facebook/presto/common/block/BlockBuilder.java [hunk 2]
```diff
@@ -118,4 +118,6 @@ default BlockBuilder appendStructureInternal(Block block, int position)
      * Creates a new block builder of the same type based on the current usage statistics of this block builder.
      */
     BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus);
+
+    void reserve(int positionCount);
```
Comment (mbasmanova) [inline presto-common/src/main/java/com/facebook/presto/common/block/BlockBuilder.java:122]: This method needs a comment.

File: presto-common/src/main/java/com/facebook/presto/common/block/BlockUtil.java [hunk 3]
```diff
@@ -266,6 +266,22 @@ else if (buffer.length < capacity) {
         return buffer;
     }
 
+    /**
+     * Returns a byte array of size capacity if the input buffer is null or smaller than the capacity.
```
Comment (mbasmanova) [inline presto-common/src/main/java/com/facebook/presto/common/block/BlockUtil.java:270]: typo: byte -> boolean

**Review by mbasmanova — CHANGES_REQUESTED** [review]
@yingsu00 The new `BlockBuilder#reserve` method is used only in one place, in GeneratedPageProjection, after a call to `BlockBuilder#newBlockBuilderLike`. It would be clearer to pass a parameter to `newBlockBuilderLike` that specifies the expected number of positions, similar to `expectedEntries` parameter in `com.facebook.presto.common.type.Type#createBlockBuilder(com.facebook.presto.common.block.BlockBuilderStatus, int)`.

```
blockBuilders = blockBuilders.stream().map(blockBuilder -> blockBuilder.newBlockBuilderLike(null, selectedPositions.size())).collect(toImmutableList());
```

**Comment by yingsu00** [comment]
> @yingsu00 The new `BlockBuilder#reserve` method is used only in one place, in GeneratedPageProjection, after a call to `BlockBuilder#newBlockBuilderLike`. It would be clearer to pass a parameter to `newBlockBuilderLike` that specifies the expected number of positions, similar to `expectedEntries` parameter in `com.facebook.presto.common.type.Type#createBlockBuilder(com.facebook.presto.common.block.BlockBuilderStatus, int)`.
> 
> ```
> blockBuilders = blockBuilders.stream().map(blockBuilder -> blockBuilder.newBlockBuilderLike(null, selectedPositions.size())).collect(toImmutableList());
> ```



> newBlockBuilderLike

@mbasmanova Thanks for reviewing. Although there is only one caller now, I'm planning to add `uncheckedWriteXXX()` and `uncheckedAppendStructure()` etc. to the BlockBuilder interface. Note that the `reserve()` currently only reserves the arrays(offsets, nulls) for the current level of BlockBuilders while not the nested ones. E.g. calling it on ArrayBlockBuilder won't reserve the element BlockBuilder's internal arrays. This is because at the time of calling `reserve` on the ArrayBlockBuilder, we don't know how many elements will be inserted into the element BlockBuilder. Thus the element BlockBuilders would still suffer from calling `ensureCapacity()` for each inserted value. The planed `uncheckedAppendStructure()` for these types will need to call `reserve` on the element BlockBuilders when it's appending a new top level row, so that the element BlockBuilders won't need to grow memory on the fly.

I can add a method `newBlockBuilderLike(null, selectedPositions.size())` which calls `reserve()`. But the fact `reserve()` being there would make the mentioned future PRs a lot easier. What do you think?

File: presto-common/src/main/java/com/facebook/presto/common/block/MapBlockBuilder.java (hunk 1 above)
Comment (yingsu00) [reply presto-common/src/main/java/com/facebook/presto/common/block/MapBlockBuilder.java:481]: @sujay-jain There is no `updateDataSize` in the MapBlockBuilder and RowBlockBuilder. I guess there were some historic reasons why they were treated differently. My understanding is the introduction of `updateDataSize` was to relieve the costly calculation in `getRetainedSizeInBytes()` but I don't have first hand data showing the benefit. So I'm just following the existing ways for now.

**Comment by mbasmanova** [comment]
@yingsu00 newBlockBuilderLike internally computes the expectedEntries and creates a new builder that pre-allocates that many entries. Passing expected number of entries to newBlockBuilderLike makes more sense, then calling it with no hint, then immediately calling `reserve`. Hence, for the use case you have in this PR, `reserve` is not the right method to use.

If `reserve` or similar method becomes necessary in the future, it can be added on first use.

**Comment by yingsu00** [comment]
> @yingsu00 newBlockBuilderLike internally computes the expectedEntries and creates a new builder that pre-allocates that many entries. 

@mbasmanova  Actually, `newBlockBuilderLike` doesn't pre-allocates any memory but just set the `initialEntryCount`:

```
@Override
    public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus)
    {
        return new Int128ArrayBlockBuilder(blockBuilderStatus, calculateBlockResetSize(positionCount));
    }

public Int128ArrayBlockBuilder(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries)
    {
        this.blockBuilderStatus = blockBuilderStatus;
        this.initialEntryCount = max(expectedEntries, 1);

        updateDataSize();
    }
```
It's up to the growCapacity() to allocate memory, but when it's uninitialized it just grow to `valueIsNull.length` which was just 1. So then whenever a new value got added, growCapacity() just grow array by 50% a time.

```
private void growCapacity()
    {
        int newSize;
        if (initialized) {
            newSize = BlockUtil.calculateNewArraySize(valueIsNull.length);
        }
        else {
            newSize = initialEntryCount;
            initialized = true;
        }

        valueIsNull = Arrays.copyOf(valueIsNull, newSize);
        values = Arrays.copyOf(values, newSize * 2);
        updateDataSize();
    }
```

If we want to pre-allocate memory, we need to do it explicitly. That's why we need `reserve()`. So we have three choices here:
1. Have reserve() in the BlockBuilders, and call it in the constructor:
```
public Int128ArrayBlockBuilder(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries)
    {
        this.blockBuilderStatus = blockBuilderStatus;
        this.initialEntryCount = max(expectedEntries, 1);
        
        reserve(expectedEntries);

        updateDataSize();
    }
```
Or
2. Change growCapacity() to
```
private void growCapacity()
    {
        int newSize = max( BlockUtil.calculateNewArraySize(valueIsNull.length), initialEntryCount);
        initialized = true;

        valueIsNull = Arrays.copyOf(valueIsNull, newSize);
        values = Arrays.copyOf(values, newSize * 2);
        updateDataSize();
    }
```
3. The third option is what this PR does. It can be changed that we add a new `public Int128ArrayBlockBuilder(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries, boolean reserveMemory)` and/or a new `newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus, int expectedEntries)`. 
```
public Int128ArrayBlockBuilder(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries, boolean reserveMemory)
    {
        this.blockBuilderStatus = blockBuilderStatus;
        this.initialEntryCount = max(expectedEntries, 1);

        //------------------------
        if (reserveMemory) {
            reserve(expectedEntries);
        }
        //--------------------

        updateDataSize();
    }

public BlockBuilder newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus, int expectedEntries)
    {
        BlockBuilder blockBuilder = new Int128ArrayBlockBuilder(blockBuilderStatus, expectedEntries, true);
    }
```
Then we can make the projection call this new constructor:
```
blockBuilders = blockBuilders.stream().map(blockBuilder -> blockBuilder.newBlockBuilderLike(null, selectedPositions.size())).collect(toImmutableList());
```

Yes true we can make these new constructors or `newBlockBuilderLike` just do what reserve() does but I think it's better to encapsulate the reserve logic in `reserve()` function for better readability and extensibility.

Both 1 and 2 would be affecting global usages. For many of the cases, we don't really know how big the BlockBuilders will be growing into. It's not very safe to do them that way. We could be seeing many BlockBuilders reserved more than the required. But for this case(projection) we do know how many elements are going to be added so it's safer to apply it just for projection for now. I tend to pick 3. What do you think?

**Comment by mbasmanova** [comment]
@yingsu00 I could be misreading the code, but here is what I understand. Int128ArrayBlockBuilder constructor doesn't allocate memory indeed, but it stores expectedEntries in initialEntryCount which is used to allocate memory in growCapacity whenever first value is added. In this case initialized is false, and growCapacity allocates enough memory to fit initialEntryCount.

**Comment by yingsu00** [comment]
@mbasmanova Thanks for reviewing the PR and bringing up the suggestion. I read the `growCapacity()` code wrong and you can disregard my last comment. I updated the PR by adding a new `newBlockBuilderLike(BlockBuilderStatus blockBuilderStatus, int expectedEntries)`. The performance benchmark shows similar improvement as before. Will you please take another look?

**Review by mbasmanova — COMMENTED** [review]
@yingsu00 Ying, is there any particular reason this change can't be simplified to just changing GeneratedPageProjection class like this?

```
public class GeneratedPageProjection
        implements PageProjection
{
    private final List<RowExpression> projections;
    private final boolean isDeterministic;
    private final InputChannels inputChannels;
    private final MethodHandle pageProjectionWorkFactory;

    private List<Type> types;

    public GeneratedPageProjection(List<RowExpression> projections, boolean isDeterministic, InputChannels inputChannels, MethodHandle pageProjectionWorkFactory)
    {
        this.projections = ImmutableList.copyOf(requireNonNull(projections, "projections is null"));
        this.isDeterministic = isDeterministic;
        this.inputChannels = requireNonNull(inputChannels, "inputChannels is null");
        this.pageProjectionWorkFactory = requireNonNull(pageProjectionWorkFactory, "pageProjectionWorkFactory is null");
        this.types = projections.stream().map(RowExpression::getType).collect(toImmutableList());
    }

    @Override
    public boolean isDeterministic()
    {
        return isDeterministic;
    }

    @Override
    public InputChannels getInputChannels()
    {
        return inputChannels;
    }

    @Override
    public Work<List<Block>> project(SqlFunctionProperties properties, DriverYieldSignal yieldSignal, Page page, SelectedPositions selectedPositions)
    {
        List<BlockBuilder> blockBuilders = types.stream()
                .map(type -> type.createBlockBuilder(null, selectedPositions.size()))
                .collect(toImmutableList());
        try {
            return (Work<List<Block>>) pageProjectionWorkFactory.invoke(blockBuilders, properties, page, selectedPositions);
        }
        catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }
```

**Comment by yingsu00** [comment]
@mbasmanova Thank you so much for the rewrite! Yes it works. I just didn't use Type.createBlockBuilder much. I have updated the PR using the way you suggested.

**Comment by mbasmanova** [comment]
@yingsu00 Thank you. I have just one more question. My read of the existing com.facebook.presto.operator.project.GeneratedPageProjection#project implementation is that first page will be processed slowly as memory will be allocated incrementally. However, second and subsequent pages will start with block builders that have expectedEntries set to the number of rows processed in the previous page. If the number of rows being processed doesn't vary much from page to page, we shouldn't see any issues after the first page. Is it the case that we are seeing memory allocations because the number of rows in pages goes up and down a lot?

**Review by mbasmanova — APPROVED** [review]

**Comment by yingsu00** [comment]
> @yingsu00 Thank you. I have just one more question. My read of the existing com.facebook.presto.operator.project.GeneratedPageProjection#project implementation is that first page will be processed slowly as memory will be allocated incrementally. However, second and subsequent pages will start with block builders that have expectedEntries set to the number of rows processed in the previous page. If the number of rows being processed doesn't vary much from page to page, we shouldn't see any issues after the first page. Is it the case that we are seeing memory allocations because the number of rows in pages goes up and down a lot?

@mbasmanova This was because the batchSize is growing from 1, 2, 4,... and BlockBuilders for previous batches is still smaller than selectedPositions.size(). Please see the following print out in `PageProcessor.processBatch()` and `LongArrayBlockBuilder.growCapacity()`:
```
processBatch batchSize=1, selectedPositions=10000, positionsBatch.size()=1
growCapacity: initialEntryCount=64, old values.length=0, new values.length=64

processBatch batchSize=2, selectedPositions=9999, positionsBatch.size()=2
growCapacity: initialEntryCount=64, old values.length=0, new values.length=64

processBatch batchSize=4, selectedPositions=9997, positionsBatch.size()=4
growCapacity: initialEntryCount=64, old values.length=0, new values.length=64

processBatch batchSize=8, selectedPositions=9993, positionsBatch.size()=8
growCapacity: initialEntryCount=64, old values.length=0, new values.length=64

processBatch batchSize=16, selectedPositions=9985, positionsBatch.size()=16
growCapacity: initialEntryCount=64, old values.length=0, new values.length=64

processBatch batchSize=32, selectedPositions=9969, positionsBatch.size()=32
growCapacity: initialEntryCount=64, old values.length=0, new values.length=64

processBatch batchSize=64, selectedPositions=9937, positionsBatch.size()=64
growCapacity: initialEntryCount=64, old values.length=0, new values.length=64

processBatch batchSize=128, selectedPositions=9873, positionsBatch.size()=128
growCapacity: initialEntryCount=80, old values.length=0, new values.length=80
growCapacity: initialEntryCount=80, old values.length=80, new values.length=120
growCapacity: initialEntryCount=80, old values.length=120, new values.length=180

processBatch batchSize=256, selectedPositions=9745, positionsBatch.size()=256
growCapacity: initialEntryCount=160, old values.length=0, new values.length=160
growCapacity: initialEntryCount=160, old values.length=160, new values.length=240
growCapacity: initialEntryCount=160, old values.length=240, new values.length=360

processBatch batchSize=512, selectedPositions=9489, positionsBatch.size()=512
growCapacity: initialEntryCount=320, old values.length=0, new values.length=320
growCapacity: initialEntryCount=320, old values.length=320, new values.length=480
growCapacity: initialEntryCount=320, old values.length=480, new values.length=720

processBatch batchSize=1024, selectedPositions=8977, positionsBatch.size()=1024
growCapacity: initialEntryCount=640, old values.length=0, new values.length=640
growCapacity: initialEntryCount=640, old values.length=640, new values.length=960
growCapacity: initialEntryCount=640, old values.length=960, new values.length=1440

processBatch batchSize=2048, selectedPositions=7953, positionsBatch.size()=2048
growCapacity: initialEntryCount=1280, old values.length=0, new values.length=1280
growCapacity: initialEntryCount=1280, old values.length=1280, new values.length=1920
growCapacity: initialEntryCount=1280, old values.length=1920, new values.length=2880

processBatch batchSize=4096, selectedPositions=5905, positionsBatch.size()=4096
growCapacity: initialEntryCount=2560, old values.length=0, new values.length=2560
growCapacity: initialEntryCount=2560, old values.length=2560, new values.length=3840
growCapacity: initialEntryCount=2560, old values.length=3840, new values.length=5760

processBatch batchSize=1809, selectedPositions=1809, positionsBatch.size()=1809
growCapacity: initialEntryCount=5120, old values.length=0, new values.length=5120
```

**Comment by mbasmanova** [comment]
@yingsu00 Thanks for explaining. I think this is because pages coming out of scan are small and grow slowly, e.g. 1, 2, 4, 8 rows per page. I understand what's going on now. I wonder if it makes sense to fix scan as well to produce large pages from the start if say all types are fixed-width.

**Comment by yingsu00** [comment]
> @yingsu00 Thanks for explaining. I think this is because pages coming out of scan are small and grow slowly, e.g. 1, 2, 4, 8 rows per page. I understand what's going on now. I wonder if it makes sense to fix scan as well to produce large pages from the start if say all types are fixed-width.

@mbasmanova Yes, when all types are fixed-width we can make the change! I was actually also thinking about the small page problem. For a more general case I was advised by @oerling not to try that because there're many pitfalls. But for all fixed-width columns case we can definitely try that. Thanks for the idea!

**Comment by yingsu00** [comment]
> @yingsu00 Thanks for explaining. I think this is because pages coming out of scan are small and grow slowly, e.g. 1, 2, 4, 8 rows per page. I understand what's going on now. I wonder if it makes sense to fix scan as well to produce large pages from the start if say all types are fixed-width.

@mbasmanova I looked at the code and found what you suggested was already implemented. If all types are fixed width, the batchSize to begin with is 1024. The reason why projection still gets small pages was because of the filters. 

However there is one improvement on the batchSize: if the included variable width columns includes just `varchar(x)` and x is predefined we can enforce stronger limit on the batchSize. However many schema use `varchar` instead of `varchar(x)`.

**Pull request merged**